x/btcstaking/keeper/power_dist_change.go

```diff
@@ -56,31 +71,56 @@ func (k Keeper) UpdatePowerDist(ctx context.Context) {
 	// to construct the new distribution
 	newDc := k.ProcessAllPowerDistUpdateEvents(ctx, dc, events, maxActiveFps)
 
-	// find newly bonded finality providers and execute the hooks
-	newBondedFinalityProviders := newDc.FindNewActiveFinalityProviders(dc)
-	for _, fp := range newBondedFinalityProviders {
-		if err := k.hooks.AfterFinalityProviderActivated(ctx, fp.BtcPk); err != nil {
-			panic(fmt.Errorf("failed to execute after finality provider %s bonded", fp.BtcPk.MarshalHex()))
-		}
-	}
-
 	// record voting power and cache for this height
-	k.recordVotingPowerAndCache(ctx, newDc)
+	k.recordVotingPowerAndCache(ctx, dc, newDc, maxActiveFps)
 	// record metrics
 	k.recordMetrics(newDc)
 }
 
-func (k Keeper) recordVotingPowerAndCache(ctx context.Context, dc *types.VotingPowerDistCache) {
+// recordVotingPowerAndCache assigns voting power to each active finality provider
+// with the following consideration:
+// 1. the fp must have timestamped pub rand
+// 2. the fp must in the top x ranked by the voting power (x is given by maxActiveFps)
+// NOTE: the previous and the new dist cache cannot be nil
+func (k Keeper) recordVotingPowerAndCache(ctx context.Context, prevDc, newDc *types.VotingPowerDistCache, maxActiveFps uint32) {
+	if prevDc == nil || newDc == nil {
+		panic("the voting power distribution cache cannot be nil")
+	}
+
 	babylonTipHeight := uint64(sdk.UnwrapSDKContext(ctx).HeaderInfo().Height)
 
-	// set voting power table for this height
-	for i := uint32(0); i < dc.NumActiveFps; i++ {
-		fp := dc.FinalityProviders[i]
+	// label fps with whether it has timestamped pub rand so that these fps
+	// will not be assigned voting power
+	for _, fp := range newDc.FinalityProviders {
+		// TODO calling HasTimestampedPubRand potentially iterates
+		// all the pub rand committed by the fp, which might slow down
+		// the process, need optimization
+		fp.IsTimestamped = k.FinalityKeeper.HasTimestampedPubRand(ctx, fp.BtcPk, babylonTipHeight)
+	}
+
+	// apply the finality provider voting power dist info to the new cache
+	// after which the cache would have active fps that are top N fps ranked
+	// by voting power with timestamped pub rand
+	newDc.ApplyActiveFinalityProviders(maxActiveFps)
+
+	// set voting power table for each active finality providers at this height
+	for i := uint32(0); i < newDc.NumActiveFps; i++ {
+		fp := newDc.FinalityProviders[i]
 		k.SetVotingPower(ctx, fp.BtcPk.MustMarshal(), babylonTipHeight, fp.TotalVotingPower)
 	}
 
+	// find newly activated finality providers and execute the hooks by comparing
+	// the previous dist cache
+	newActivatedFinalityProviders := newDc.FindNewActiveFinalityProviders(prevDc)
```

Comment:
for the case where `len(events) == 0` and `dc != nil` that we do the deep copy, it will never have a set of `newActivatedFinalityProviders` ?

Reply:
I thought the same but turned out there is an issue if we don't take care of this. So, even if no new events, some fps could have newly timestamped pub rand and become a new activated fp. For example, we only have one fp and one delegation to it at height `100` but it does not have timestamped pub rand. When it comes to height `101`, the fp's pub rand commit is timestamped, so it will be assigned voting power for `101`, thus it becomes activated

Reply:
got it, thanks for explaining! 
In this case, we don't have a way to avoid the deep copy :/